[2단계 - 장바구니 기능] 도이(유도영) 미션 제출합니다.

yoondgu opened the pull request:
안녕하세요 빙봉!! 디엠 드렸지만 더욱 더 늦게 요청드립니다.. 죄송합니다 😢!!

1단계 추가 코멘트 반영사항, 2단계 요구사항을 README.md에 정리했습니다.
그리고 1단계, 2단계 통틀어 API 명세 또한 작성해보았어요.

### 이메일, 비밀번호 형식 검증은 클라이언트 책임으로 판단
Product 도메인 객체에서 검증을 수행했던 것처럼 해주어야 할까? 생각했지만 
이번 미션에서는 이메일, 비밀번호가 인증 목적으로만 사용되기 때문에 핵심 비즈니스 로직에서 필요한 부분은 아니라고 판단했습니다.
만약 회원가입 로직이 있었다면 넣었겠지만, 인증을 거친 후에는 비즈니스 로직에서 사용자의 Id만 사용하기 때문입니다.
그래서 클라이언트에서 형식 검증이 완료된 뒤 값을 받는다고 생각하고 도메인에서 형식 검증 로직은 넣지 않았습니다.

### Interceptor, ArgumentResolver 로 인증 책임 분리
- 미션 요구사항에서 주어진 대로 Basic 인증을 사용하였습니다.
- Interceptor, ArgumentResolver 모두 사용해서 인증 과정의 책임을 조금 더 분리해보고자 했습니다.
    - AuthenticationInterceptor는 헤더 값 검증 후, 컨트롤러에서 email,password를 모두 알 필요 없도록 사용자의 id를 request 객체의 속성에 저장합니다.
    - AuthenticationPrincipalArgumentResolver 는 로그인 사용자의 id를 받아와 서비스에서 사용할 수 있도록 합니다.
    
### Dao 제너릭 인터페이스의 재사용 -> 추상클래스로 도메인 별 Dao 정의 -> 캡슐화를 위한 패키지 구조 변경
 - 지난 번 코멘트 주신 대로 Dao 인터페이스를 제너릭으로 만들어보았는데요. CRUD는 대부분 비슷한 형식의 메서드 시그니처를 가지기에, 이 방식을 계속 유지해보고 싶었어요.
 - 하지만 각 Dao마다 어떤 메서드는 사용하고, 어떤 메서드는 사용하지 않고, 해당 테이블에서만 특화된 메서드가 있는 등 생각보다 활용하기가 쉽지 않았습니다.
 - 이를 해결하는 방법을 고민하다가, 중간에 추상클래스 계층을 두고 안쓰는 메서드는 final로 재정의를 막은 뒤 지원하지 않는 기능이라는 예외를 던지도록 만들어보았습니다. 인터페이스 구현체가 바로 예외를 던지지 않게 만든 이유는, H2ProductDao, OracleProductDao.. 등 벤더사 별로 갈아끼울 수 있도록 하되 각 Dao들이 사용하는 메서드/사용하지 않는 메서드를 똑같이 알게 하기 위함이었습니다.
 1. 어떤 DAO든 사용하는 CRUD 메서드명, 시그니처 통일을 위한 인터페이스 `Dao<T>`
 2. XXX 테이블에 대한 DAO의 지원 메서드를 정의하는 추상클래스 `XXXDao extends Dao<XXX>`
 3. XXX 테이블에 대한 DAO 구현체 `H2XXXDao, MySQLXXXDao`
 - 하지만 추상클래스의 남용을 막고자 package-private 으로 두어야 했고, 동시에 Service는 이를 알아야 하므로 패키지 구조를 도메인 별로 나누게 되었습니다.
 -  이번 미션 요구사항에서 객체와 테이블 그리고 패키지 구조는 자유롭게 설계할 수 있다고 하여서, 학습 목적으로 적용해보았는데요. 사실 이렇게 구조를 바꾸어 생긴 단점이 더 많은 것 같습니다. 
 1. 지원하는 메서드인 줄 알고 사용했다가 예외가 터질 수 있다 (예측 불가능한 코드)
 2. 프로젝트 규모에 비해 오버엔지니어링이다
 3. 하나의 Dao에 여러 계층이 나누어져 있어 코드를 읽고 파악하기 불편하다

다시 구조를 되돌릴까 싶지만, 그 전에 빙봉은 어떻게 생각하시는지 여쭙고 싶어요.!! 

+ DB 실행 환경 관련 어색한 설정이 있는데 아직 해결에 어려움을 겪고 있어 [코멘트](https://github.com/woowacourse/jwp-shopping-cart/pull/337#discussion_r1186655949) 남겼습니다..ㅎㅎ

이번 리뷰도 잘 부탁드립니다!! 감사드려요 🐘🐘🐘

yoondgu committed:
docs: 피드백 목록 1단계 추가 코멘트 반영

yoondgu committed:
refactor: 서비스에 불필요한 트랜잭션 어노테이션 삭제

yoondgu committed:
refactor: 뷰 컨트롤러 ModelAndView 사용 방식으로 변경

yoondgu committed:
refactor: 예측하지 못한 예외에 대한 Handler 추가

yoondgu committed:
refactor: RestController 복수형으로 자원 표기하도록 URI 변경

yoondgu committed:
refactor: 예외메시지 수정, 관련 TODO 추가

yoondgu committed:
docs: API 명세 작성

yoondgu committed:
feat: 사용자 설정 페이지 연동, 사용자 객체 구현

yoondgu committed:
feat: 사용자 테이블 설계, 초기화 설정

yoondgu committed:
feat: 사용자 DB 조회 구현

yoondgu committed:
feat: 사용자 설정 페이지, DB 연동 완료

yoondgu committed:
feat: 장바구니 페이지 연동

yoondgu committed:
docs: API 명세 body name 오기 수정

yoondgu committed:
docs: 장바구니 아이템 테이블 설계, 장바구니 관리 API 설계

yoondgu committed:
feat: 장바구니 아이템 객체 구현

- CartItem이 Product 객체 참조하게 하고, 응답용 DTO는 따로 만들지 검토 필요

yoondgu committed:
refactor: Dao 인터페이스 구조 단순화

- 일단 하나의 Dao 인터페이스를 구현하고, 사용하지 않는 메서드는 예외 던지도록 함
- 인터페이스 외 추가 메서드 정의에 대해서 추후 검토 필요
- 도메인 별 패키지 구조 전환 후 추상 클래스 활용 추후 검토 필요

yoondgu committed:
refactor: 패키지 구조 도메인 별로 변경, 도메인 별 Dao 추상 클래스 정의

- 기본적인 CRUD 관련 메서드 표준을 제공하는 Dao 인터페이스를 두되
- 도메인 별로 사용하지 않는 메서드/추가할 메서드를 정의하기 위해 추상 클래스 정의
- 추상 클래스의 접근 제어자로 의도치 않은 상속 방지하기 위해 패키지 구조 변경함.
- 학습 목적으로 진행하였으나 각 도메인 별 Dao의 인터페이스를 다르게 가지는 것이 낫다고 판단되면 다시 수정 예정.

yoondgu committed:
feat: 이메일로 사용자 조회 구현

yoondgu committed:
feat: 장바구니 아이템 조회 uri 매핑

yoondgu committed:
chore: 장바구니 아이템 DDL 추가, 상품 더미데이터 DML 추가

yoondgu committed:
feat: 장바구니 아이템 DAO 구현

yoondgu committed:
feat: 장바구니 아이템 등록 기능 구현

yoondgu committed:
fix: 장바구니 페이지 변수명 오기 수정

yoondgu committed:
feat: 장바구니 아이템 조회 기능 구현

yoondgu committed:
feat: 장바구니 아이템 삭제 기능 구현

yoondgu committed:
feat: 장바구니 아이템 외래키 연쇄 삭제 제약조건 추가

yoondgu committed:
refactor: 장바구니 아이템 컨트롤러 메서드 순서 수정

yoondgu committed:
refactor: 사용자 인증 절차 중복 제거

yoondgu committed:
refactor: 인증 실패 커스텀 예외 정의, 응답코드 설정

yoondgu committed:
refactor: 장바구니 아이템 삭제 시 사용자 검증 로직 추가

yoondgu committed:
refactor: 요청객체 로그인 사용자 ID 속성명 상수화

yoondgu committed:
refactor: DTO Null 검증 예외 메시지 구체화

yoondgu committed:
refactor: 컨트롤러 불필요한 의존성 삭제

yoondgu committed:
refactor: 테스트 코드 상수 처리

yoondgu committed:
refactor: 테스트 픽스쳐 적용

- 테스트 DB 초기데이터와 일치하도록 설정

yoondgu committed:
test: 장바구니 아이템 삭제 시 사용자 검증 테스트 추가

yoondgu committed:
fix: 장바구니 화면 요청 시 인증 인터셉터 실행하는 문제 수정

- 인증 인터셉터 잘못된 경로 설정 수정

yoondgu committed:
docs: CART_ITEMS_API 응답 실패 메시지 작성

yoondgu committed:
build: 예외 관련 패키지 분리

yoondgu committed:
style: 불필요한 주석 삭제

yoondgu commented on src/main/resources/schema.sql:
```diff
@@ -1,7 +1,33 @@
-    price INT NOT NULL,
+    price     INT             NOT NULL,
     PRIMARY KEY (id)
 );
```
실제 서비스 환경과 비슷하게 하고 싶어서 Embedded DB를 사용했던 건데,
 테스트 편의를 위해 넣는 INSERT문(products)을 사용하는 상황에서
 DROP 이나 TRUNCATE을 사용하지 않고도
기존 데이터가 있으면 넣지 않고 아니면 넣게 하는 법을 아직 찾지 못해서
 결국 DDL에서 인메모리처럼 프로그램 실행 시 마다 테이블을 새로 만들고 있습니다 😢😢

aegis1920 replied:
- 제가 리뷰했던 다른 리뷰이와 같은 이슈를 겪고 있군요ㅎㅎ 여기 답변이 있으니 한 번 적용해보시죠! ([링크](https://github.com/woowacourse/jwp-shopping-cart/pull/189#discussion_r1181579031))
- 추가로 이 내용도 읽어보시면 좋을 것 같습니다 ([링크](https://github.com/woowacourse/jwp-shopping-cart/pull/189#discussion_r1181595490))

yoondgu replied:
INSERT IGNORE가 있는지 처음 알았어요 .. 👍👍
감사합니다 ㅎㅎㅎ
추가로 주신 링크 내용도 도움이 많이 되었어요!

aegis1920 commented on docs/CART_ITEMS_API.md:
REST API 명세 멋집니다! 👍 이런 문서화는 테스트코드 기반 혹은 어노테이션 기반으로 문서화 관리가 가능한데요. 추후 프로젝트하실 때 적용해보시는 것도 좋을 것 같습니다! (RestDocs, Swagger 등등..)

yoondgu replied:
우와 이런 게 있군요!! 다음 미션에서부터 사용해보아야겠습니다. 알려주셔서 감사합니다 ㅎㅎㅎ

aegis1920 commented on src/test/java/cart/domain/cartitem/H2CartItemDaoTest.java:
```diff
@@ -0,0 +1,73 @@
+
+@JdbcTest
+@AutoConfigureTestDatabase(replace = Replace.NONE)
+class H2CartItemDaoTest {
```
전체 테스트를 돌려보면 해당 테스트 실패하는 케이스가 있네요! 테스트는 독립적이어야 합니다! 확인 부탁드려요
<img width="687" alt="image" src="https://user-images.githubusercontent.com/22311557/236657374-0965c403-3947-476c-ace7-e67124853f22.png">
```
PreparedStatementCallback; Referential integrity constraint violation: "CONSTRAINT_D: PUBLIC.CART_ITEMS FOREIGN KEY(MEMBER_ID) REFERENCES PUBLIC.MEMBERS(ID) (CAST(1 AS BIGINT))"; SQL statement:
INSERT INTO cart_items (MEMBER_ID, PRODUCT_ID) VALUES(?, ?) [23506-214]; nested exception is org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: Referential integrity constraint violation: "CONSTRAINT_D: PUBLIC.CART_ITEMS FOREIGN KEY(MEMBER_ID) REFERENCES PUBLIC.MEMBERS(ID) (CAST(1 AS BIGINT))"; SQL statement:
```

yoondgu replied:
앗 전체 테스트를 돌렸을 때, `H2CartItemDaoTest`가 `H2ProductDaoTest`보다 늦게 실행될 경우 실패한다는 것을 알게 되었습니다.
리뷰 요청 전에 돌렸을 때는 운좋게 순서가 달라서 모두 성공했던 거네요 😭😭

실패 케이스를 디버깅해보니 `CartItemDaoTest`에서는 members DB 데이터가 존재하지 않아 위와 같은 예외가 발생한 것이었습니다.

1. `H2ProductDaoTest` 실행 전 `H2CartItemDaoTest` 실행
members 테이블에 초기 데이터가 들어가있음
![image](https://user-images.githubusercontent.com/97426362/236671212-a6335f3f-1468-4628-a61a-c9633ff97d1e.png)

2. `H2ProductDaoTest` 실행 후 `H2CartItemDaoTest` 실행
members 테이블에 초기 데이터가 들어가있지 않음
![image](https://user-images.githubusercontent.com/97426362/236676299-798ca14c-c9f5-4653-bc57-b66ec0b15568.png)


그리고 다시 코드를 살펴보니, 제가 `H2ProductDaoTest`에 불필요한 @SqlGroup, @Sql 어노테이션을 적용하고 있었습니다. 
(설정하지 않아도 파일 경로 상 schema.sql, data.sql을 실행함)
해당 어노테이션들을 지우니 `H2ProductDaoTest` 이후에 `H2CartItemDaoTest`를 실행해도 정상 작동함을 확인하고 이를 수정하였습니다.
하지만 @SqlGroup과의 인과관계는 아직 알지 못해서 `SettingsCreator`, `SqlScriptsTestExecutionListener`에 브레이크포인트를 걸고 디버깅해보았으나 아직 잘 파악이 되지 않습니다..!
추측으로는 `H2ProductDaoTest`의 @SqlGroup에서 초기화 하는 과정에서 어떤 설정이 바뀌어서, 그 다음 `H2CartItemDaoTest`를 실행할 때는 data.sql을 실행하지 않는 것 같은데 잘 모르겠습니다..

yoondgu replied:
추가로 [DB 초기화 설정](https://github.com/woowacourse/jwp-shopping-cart/pull/337#discussion_r1186787097) 관련 코멘트 주신 부분을 참고하여 로컬 DB면 로컬 DB 답게 테이블을 지울 필요 없이 수정해보았는데,
테스트 DB에서도 같은 방식을 사용하니 테스트 간 독립적이기 어렵다는 것을 깨달았습니다.
(truncate을 해주되 auto_increment 값을 고려해야 하는 등 설정을 조금만 잘못해도 테스트가 여기저기 터지는데 디버깅이 어려웠습니다..)
결국 테스트 독립성을 위해 매번 테이블 초기화 설정을 해줘야 함 => 차라리 테스트 DB는 인메모리를 쓰는 게 낫다고 판단해 이를 수정하였습니다..!
사실 인메모리를 쓰면 위 코멘트에서 말씀드린 문제도 고려할 필요 없기도 하고요.
현업에서는 테스트용 DB로 주로 로컬에 저장되는 DB /. 인메모리 중 어떤 것을 사용하나요..??

aegis1920 replied:
> 추측으로는 H2ProductDaoTest의 @SqlGroup에서 초기화 하는 과정에서 어떤 설정이 바뀌어서, 그 다음 H2CartItemDaoTest를 실행할 때는 data.sql을 실행하지 않는 것 같은데 잘 모르겠습니다..

이 부분은 자세히 보지 않은 터라 잘 모르겠네요ㅎㅎ 그리고 지금 `@SqlGroup`넣고 로컬 DB로 경로를 바꾼 후 테스트를 돌려봐도 깨지지 않고 있습니다. 그렇게 중요한 부분은 아니니 추후 문제가 발생했을 때 다시 살펴봐도 될 것 같아요!

> 사실 인메모리를 쓰면 위 코멘트에서 말씀드린 문제도 고려할 필요 없기도 하고요. 현업에서는 테스트용 DB로 주로 로컬에 저장되는 DB /. 인메모리 중 어떤 것을 사용하나요..??

실무에서는 테이블 간 관계가 매우 복잡하고 데이터 양도 많아서요. 그래서 인메모리로 따로 올리기에는 신경쓸 부분이 많아 쓰지 않습니다. 단위테스트같은 경우는 DB에서 가져오는 데이터들을 fake 객체로 직접 만들어 테스트하고 통합테스트할 때는 개발환경의 DB를 dump한 image를 Docker로 로컬에 띄워서 테스트하고 있습니다! (물론 팀마다 다를 수 있습니다 🙂 )

aegis1920 commented on src/main/java/cart/controller/CartItemController.java:
```diff
@@ -0,0 +1,44 @@
+        return cartItemService.findAllByMemberId(loginMemberId);
+    }
+
+    @ResponseStatus(HttpStatus.CREATED)
```
ResponseEntity를 사용해보는 건 어떨까요? 확장성 측면에서 특정 헤더가 추가될 경우 좀 더 깔끔하게 추가할 수 있을 것 같아요.

aegis1920 commented on src/main/java/cart/infratstructure/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,42 @@
+        Member member = memberService.findByEmail(email)
+                .orElseThrow(() -> new AuthenticationException("잘못된 사용자 정보입니다"));
+        if (!Objects.equals(password, member.getPassword())) {
+            throw new AuthenticationException("잘못된 사용자 정보입니다");
```
에러가 발생했을 때 클라이언트쪽에서 간단하게 alert해줘도 좋을 것 같아요!

aegis1920 commented on src/main/java/cart/controller/CartItemController.java:
```diff
@@ -0,0 +1,44 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/cartitems")
```
URL에 구분자를 추가하는 게 좋을 것 같아요. `_`가 좋을까요? 아니면 `-`가 좋을까요?

yoondgu replied:
_보다 -가 가독성이 더 좋은 것 같아서 -를 사용하도록 하겠습니다!
또, 절대적인 규칙은 아니지만 일반적으로도 가독성을 이유로 -를 더 많이 쓰는 것 같아서, 관례를 따르는 것도 좋다고 생각합니다.

aegis1920 replied:
가독성보다는 검색 엔진에 더 유리하기 때문인데요. 이쪽 링크를 읽어보셔도 좋을 것 같습니다! ([링크](https://developers.google.com/search/docs/crawling-indexing/url-structure?hl=en&visit_id=638192436751061801-3816480568&rd=1))

aegis1920 commented on src/main/java/cart/dto/AuthInfo.java:
```diff
@@ -0,0 +1,19 @@
+package cart.dto;
+
+public class AuthInfo {
```
> 만약 회원가입 로직이 있었다면 넣었겠지만, 인증을 거친 후에는 비즈니스 로직에서 사용자의 Id만 사용하기 때문입니다. 그래서 클라이언트에서 형식 검증이 완료된 뒤 값을 받는다고 생각하고 도메인에서 형식 검증 로직은 넣지 않았습니다.

회원가입 로직이 없다고 하더라도 (인증 작업만 있다고 하더라도) 검증이 필요하지 않을까요? 검증 로직이 없다면 사용자가 반복해서 에러를 낼 수 있는 상황으로 보이는데 이 자체가 문제가 되지 않을까요?

yoondgu replied:
이메일 형식부터 맞지 않아 당연히 인증이 실패할 사용자정보로 요청을 보낼 경우,
검증 로직이 있다면 영속 계층까지 내려가지 않고도 예외를 던질 수 있어 좋지 않은 것 같네요..!!
검증로직을 추가하도록 하겠습니다 !

yoondgu replied:
그런데 궁금증이 생겼습니다. 😯

현재 코드에서 이메일, 비밀번호는 아래와 같은 흐름으로 사용됩니다.
1. 브라우저에서 헤더에 Authorization 정보를 넣어 요청
2. Interceptor에서 이를 추출하여 Email, Password 값을 검증
(memberService.findByEmail() 메서드 사용하여 Member 조회)
3. 검증된 Member의 id값만을 Controller에서 받아옴
4. 비즈니스 로직에서 memberId를 사용해 DB 접근 로직 수행

따라서 잘못된 형식의 Email, Password라도 findByEmail()에서 조회가 안될테니 (2)에서 요청 처리가 끝날 것이고,
Member 객체는 DB에서 조회해온 회원 정보를 나타내는 데에만 사용되므로
Member 클래스에서 Email, Password 검증 로직을 만들어두어도 그것이 사용될 일이 없습니다.

따라서 Member 클래스가 아니라 (2)에서 Interceptor가 추출한 Email, Password에 대한 형식 검증을 먼저 하고,
그 다음에야 DB 조회를 통한 검증을 하는 방법으로 진행하려고 합니다.

그런데 도메인에 대한 검증을 Interceptor에서 하는 것이 조금 어색하게 느껴집니다.

**그리고 만약 이 서비스에서 지정한 Email, Password에 대한 형식 정책이 중간에 바뀔 수 있다면,**
DB에 저장된 회원 정보 중 어떤 것은 지금 형식 정책과는 맞지 않을 수도 있지 않을까요?
이 경우에는 인증이 성공해야 함에도 불구하고 형식 검증에서 실패를 할 수도 있지 않을까요..??

제가 빙봉께서 주신 질문을 잘 이해하고 한 고민인지, 이에 대해 어떻게 생각하시는지 궁금합니다!!

aegis1920 replied:
> 그런데 도메인에 대한 검증을 Interceptor에서 하는 것이 조금 어색하게 느껴집니다.

말씀해주신 도메인은 AuthInfo를 말씀해주신 거겠죠? AuthInfo는 생성될 때 검증되는데 Interceptor에서 해도 어색하지 않다고 생각하는데요. 이유는 Service에서 도메인을 생성할 때 도메인 내에서 검증되는 것과 같다고 생각해서요. 어떤 부분이 어색하게 느껴지시는 지 궁금하네요 🙂 

> 그리고 만약 이 서비스에서 지정한 Email, Password에 대한 형식 정책이 중간에 바뀔 수 있다면, DB에 저장된 회원 정보 중 어떤 것은 지금 형식 정책과는 맞지 않을 수도 있지 않을까요? 이 경우에는 인증이 성공해야 함에도 불구하고 형식 검증에서 실패를 할 수도 있지 않을까요..??

형식 정책이 중간에 바뀐다면 당연히 바뀐 정책으로 코드도 변경해줘야겠죠? DB에 이미 저장된 회원 정보라면 그 회원 정보 또한 바뀐 정책으로 마이그레이션을 해줘야 합니다! 회원정보같은 정책이 바뀌면 그대로 둘 수 없어요!

> 제가 빙봉께서 주신 질문을 잘 이해하고 한 고민인지, 이에 대해 어떻게 생각하시는지 궁금합니다!!

네네! 잘 이해하셨고 질문도 좋았습니다 👍

aegis1920 commented on src/main/java/cart/infratstructure/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,35 @@
+import org.apache.tomcat.util.codec.binary.Base64;
+
+public class BasicAuthorizationExtractor {
+    private static String AUTHORIZATION = "Authorization";
```
- final 확인 부탁드려요!
- Authorization 자체를 의미하기 때문에 굳이 상수로 빼지 않아도 될 것 같긴 하네요.

aegis1920 commented on src/main/java/cart/infratstructure/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,35 @@
+
+public class BasicAuthorizationExtractor {
+    private static String AUTHORIZATION = "Authorization";
+    private static final String BASIC_TYPE = "Basic";
```
아래 로직에서 lowerCase를 수행하는데 `"basic"`으로 만들면 되지 않을까요?

aegis1920 commented on src/main/java/cart/infratstructure/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,35 @@
+            String password = credentials[1];
+
+            return new AuthInfo(email, password);
+        }
```
함수 분리 부탁드려요!

aegis1920 commented on src/main/java/cart/infratstructure/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,35 @@
+
+        if (header == null) {
+            return null;
+        }
```
아래에서 똑같이 return null이라면 둘이 합쳐도 좋을 것 같습니다.

```
if (header == null || !header.toLowerCase().startsWith(BASIC_TYPE.toLowerCase())) {
    return null;
}
```

aegis1920 commented on src/main/java/cart/infratstructure/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,35 @@
+            return null;
+        }
+
+        if ((header.toLowerCase().startsWith(BASIC_TYPE.toLowerCase()))) {
```
```suggestion
        if (header.toLowerCase().startsWith(BASIC_TYPE.toLowerCase())) {
```

aegis1920 commented on src/main/java/cart/infratstructure/BasicAuthorizationExtractor.java:
```diff
@@ -0,0 +1,35 @@
+
+            return new AuthInfo(email, password);
+        }
+
```
공백이 있군요!

yoondgu replied:
구현을 급하게 하다보니 기본적인 부분들을 많이 신경을 못썼네요 ㅜㅜ 확인해주셔서 감사합니다!! 모두 수정하겠습니다.

aegis1920 commented on src/main/java/cart/infratstructure/AuthenticationPrincipal.java:
```diff
@@ -0,0 +1,12 @@
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface AuthenticationPrincipal {
+    boolean required() default true;
```
필요한 로직인가요?

yoondgu replied:
기존 학습 테스트에서 주어진 코드를 활용하는 과정에서 이런 부분을 꼼꼼하게 살피지 못했습니다 ㅜㅜ
현재 필요하지 않은 로직이므로 삭제하도록 하겠습니다!

aegis1920 commented on src/main/java/cart/infratstructure/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,42 @@
+        }
+
+        request.setAttribute(LOGIN_MEMBER_ID.name(), member.getId());
+        return HandlerInterceptor.super.preHandle(request, response, handler);
```
super의 preHandle을 호출하신 이유가 있을까요?

yoondgu replied:
컨트롤러 요청 처리를 마저 진행하도록 return true를 해주면 되는데,
intelliJ override 자동완성에서 제공된 코드를 그대로 사용해버렸습니다 🥲 수정할게요!

aegis1920 commented on src/main/java/cart/infratstructure/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,42 @@
+
+public class AuthenticationInterceptor implements HandlerInterceptor {
+
+    private final BasicAuthorizationExtractor basicAuthorizationExtractor = new BasicAuthorizationExtractor();
```
Extractor를 생성해서 할당해주신 이유를 알 수 있을까요? (Bean으로 만들어도 되지 않을까 싶었는데... Request를 받아서 추출하는 역할만 하면 되니 Bean으로 굳이 만들지 않아도 될 것 같기도 하고요.. 도이의 생각이 궁금합니다!)

yoondgu replied:
사실 깊게 고민을 하지 못하고 진행한 부분인 것 같아요.
하지만 생각해보니, 
1. 어플리케이션 생명주기에 따라 단 하나를 생성하고 재사용하는 객체
2. 의존성 주입으로 갈아끼우는 일이 필요한 객체
는 빈으로 등록하는 게 좋은 것 같습니다.

현재는 (1)에만 해당하지만, 만약 인증 종류가 바뀐다면 BasicAuthroizationExtractor가 아닌 다른 AuthorizationExtractor를 사용하도록 이 클래스의 코드 수정을 해줘야 할 것입니다. (OCP 위반)
이런 문제를 방지하기 위해서는, AuthorizationExtractor 인터페이스를 정의하고 이를 구현하는 BasicAuthroizationExtractor를 빈 등록한 뒤 의존성 주입받는 것이 좋을 것 같습니다.
오버엔지니어링 보다는 유지보수성을 향상하는 일에 가까운 것 같아서 말씀드린 방향으로 수정하겠습니다!

aegis1920 replied:
인증 종류가 바뀔 수 있다는 맥락에서 Bean으로 설정하는 것도 좋아보이네요. 좋습니다! 👍
Bean이 많아지면 Bean 초기화 속도도 늘어나고 관리도 힘들어지니 유의해주세요!

aegis1920 commented on src/main/java/cart/infratstructure/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,42 @@
+            throw new AuthenticationException("잘못된 사용자 정보입니다");
+        }
+
+        request.setAttribute(LOGIN_MEMBER_ID.name(), member.getId());
```
> AuthenticationInterceptor는 헤더 값 검증 후, 컨트롤러에서 email,password를 모두 알 필요 없도록 사용자의 id를 request 객체의 속성에 저장합니다.

좋은 방법이네요 👍 👍

aegis1920 commented on src/main/java/cart/domain/member/MemberDao.java:
```diff
@@ -0,0 +1,34 @@
+import cart.dao.Dao;
+import java.util.Optional;
+
+abstract class MemberDao implements Dao<Member> {
```
> 하지만 추상클래스의 남용을 막고자 package-private 으로 두어야 했고, 동시에 Service는 이를 알아야 하므로 패키지 구조를 도메인 별로 나누게 되었습니다.
이번 미션 요구사항에서 객체와 테이블 그리고 패키지 구조는 자유롭게 설계할 수 있다고 하여서, 학습 목적으로 적용해보았는데요. 사실 이렇게 구조를 바꾸어 생긴 단점이 더 많은 것 같습니다.

패키지 구조를 바뀐 점으로 인해 많은 걸 깨닫고 계시군요 👍 👍 

> 지원하는 메서드인 줄 알고 사용했다가 예외가 터질 수 있다 (예측 불가능한 코드) 프로젝트 규모에 비해 오버엔지니어링이다 하나의 Dao에 여러 계층이 나누어져 있어 코드를 읽고 파악하기 불편하다 다시 구조를 되돌릴까 싶지만, 그 전에 빙봉은 어떻게 생각하시는지 여쭙고 싶어요.!!

- 지금 테이블과 벤더사 모두 통일되도록 해야해서 구현하시기 더 어려울거라 생각합니다. 기대했던 바는 DAO 인터페이스에는 공통으로 사용되는 코드를 만들고 개별적으로 필요한 코드는 그에 따른 인터페이스를 다시 추가해서 만드는 것을 생각하긴 했습니다. 
  - `MemberDao` Class <- `MemberDao` interface <- `Dao<Member>` interface
  - 왜냐하면 JpaRepository가 이런 형식으로 구성되어있거든요. 지금 이런 걸 경험해보고 추후 JPA를 사용하실 때 좀 더 이해하기 쉽지 않을까 생각했습니다 🙂 
- 결국 직접 경험해보고 느끼신 게 중요하거든요. 말씀해주신 단점 모두 공감가지만 지금 구현도 잘 하셨다고 생각합니다 👍 일단 이번 미션에서는 패키지 변경에 대해 단점을 느끼셨다는 것만으로도 충분하다고 생각하고 굳이 바꾸진 않으셔도 될 것 같아요. 여기서 시간을 더 쓰시기 보단 다른 걸 학습해보세요!

yoondgu replied:
감사합니다!! Dao 클래스에서 공통되는 부분을 추상화할 생각을 못해봤는데, 빙봉 덕분에 도움이 되는, 재밌는 경험을 할 수 있었던 것 같아요.
빨리 JPA를 써보고 싶어졌습니다.. ㅋㅋ
말씀해주신 것처럼 벤더사도 통일되게 하려고 하다보니 어려웠던 것 같아요. `MemberDao` 부분을 인터페이스로 하려다가, 사용하지 않는 기능에 대해 예외를 던지는 일을 `MemberDao`의 구현체에게 넘겨주면 벤더사 별 통일을 강제할 수 없어 추상클래스로 진행했습니다. 하지만 이 문제만 아니라면 인터페이스를 쓰는 것이 좋아보여요..!! JpaRepository에 대해 더 공부해봐야겠습니다 ㅎㅎ 감사합니다!

aegis1920 commented on src/main/resources/schema.sql:
```diff
@@ -1,7 +1,33 @@
+    id         BIGINT UNSIGNED NOT NULL AUTO_INCREMENT,
+    member_id  BIGINT UNSIGNED NOT NULL,
+    product_id BIGINT UNSIGNED NOT NULL,
+    FOREIGN KEY (member_id) REFERENCES members (id) ON DELETE CASCADE,
```
외래키를 사용하신 이유가 있을까요? 어떤 장점과 어떤 단점이 있을까요?

yoondgu replied:
테이블 간 중복 정보를 줄이기 위하여 외래키를 사용했습니다!
중복을 줄이고, 데이터의 무결성을 지킬 수 있다는 장점이 있습니다.
하지만 성능적 단점, 테이블 구조 변경 시 제약이 있다는 점, 테스트 시에도 데이터 관리가 어렵다는 단점이 있습니다.

또 제가 느낀 단점으로는,
id를 기준으로 식별하다보니 도메인 객체를 테이블에 맞는 엔티티로 사용할 경우, 참조 방식의 연관관계를 맺을 수 없어 객체를 객체답게 사용하지 못하는 점이 불편했습니다.
현재도 구조를 단순하게 가져가고 싶어서 도메인 객체를 DB용 엔티티로 사용하다보니
CartItem 클래스가 Member, Product를 참조하는 방식으로 만들 수 없었습니다.
하지만 이 점은 패러다임이 불일치하여 발생하는, 피할 수 없는 문제라 도메인 객체, 순수 DB용 엔티티 객체를 분리하여 변환해 사용하도록 하면 되는 것 같네요!

aegis1920 replied:
- 맞습니다! 외래키를 적용하면 성능 저하 및 여러 작업(마이그레이션 혹은 삽입, 삭제 등등..) 시 고려해야할 점들이 늘어나게 됩니다. 이런 불편으로 실무에서도 외래키는 가급적 사용하지 않습니다.
- [github이 외래키를 쓰지 않는 이유](https://github.com/github/gh-ost/issues/331)도 읽어보시면 좋을 것 같아요!

aegis1920 reviewed (changes requested):
안녕하세요 도이! 빙봉입니다 🙂 

질문에 대한 답변 및 코멘트 남겼으니 확인해주세요!

궁금한 점은 언제든 편하게 DM 주세요~

yoondgu committed:
fix: 사용자 아이디, 상품 아이디 조회 순서 오류 수정

yoondgu committed:
test: 일부 테스트 실행 순서에 따라 실패 오류 수정

- ProductItemDaoTest에 @SqlGroup 어노테이션 적용 시, CartItemDaoTest가 나중에 실행되면 실패하는 오류 발견.
- 해당 상황 시 CartItemDaoTest에서는 data.sql이 실행되지 않아 members DB 데이터가 존재하지 않음.
- 이에 따라 JdbcSQLIntegrityConstraintViolationException 예외 발생함.
- @SqlGroup 어노테이션 삭제 (불필요한 적용이었음)

yoondgu committed:
build: DB 초기화 전략 수정

- 기존 초기화 전략은 로컬 DB를 사용함에도 인메모리처럼 쓰고 있음.
- 초기 데이터 DML에 INSERT IGNORE~ 사용하고 DDL에서는 테이블 다시 생성할 필요 없도록 수정함.
- 테스트 격리 문제는 서로 다른 DAO 간 의존을 없애기 위해 인메모리 DB로 변경함.
- 클래스 내에서 메서드 간 테스트 격리가 필요한 @SpringBootTest에서는 afterEach로 직접 해결함.

yoondgu committed:
docs: 2단계 피드백 및 리팩터링 목록 추가

yoondgu committed:
refactor: ResController 확장성 고려하여 어노테이션 대신 ResponseEntity 사용

yoondgu committed:
feat: 에러 발생 시 클라이언트 쪽 alert 띄우기

yoondgu committed:
refactor: /cartitems URL 구분자 추가

yoondgu committed:
docs: 2단계 피드백 목록 누락 추가

yoondgu committed:
refactor: AuthenticationInterceptor 반환값 수정

yoondgu committed:
refactor: AuthenticationPrincipal 불필요한 메서드 삭제

yoondgu committed:
refactor: 인증 Extractor Bean 의존성 주입으로 변경

yoondgu committed:
refactor: 인증 관련 클래스 클린 코드를 위한 수정

- final 키워드 누락 수정
- 불필요한 상수 선언, 소문자 변환 등 수정
- 조건문 정리
- 함수 분리
- 공백 삭제

yoondgu committed:
fix: AuthorizationExtractor 의존성 주입 설정

yoondgu committed:
refactor: AuthenticationInterceptor 함수 분리

yoondgu committed:
refactor: 이메일, 비밀번호 검증 로직 추가

yoondgu committed:
refactor: json 주고받는 데이터 carmel case로 통일

yoondgu committed:
test: BasicAuthorizationExtractorTest 의존성 주입 실패

yoondgu commented on src/test/java/cart/infratstructure/BasicAuthorizationExtractorTest.java:
```diff
@@ -0,0 +1,40 @@
+//class BasicAuthorizationExtractorTest {
+//
+//    @Autowired
+//    private BasicAuthorizationExtractor basicAuthorizationExtractor; // 의존성 주입 되지 않아 null 할당
```
BasicAuthorizationExtractor에 대한 테스트를 작성하고 제출하려고 했는데,
의존성 주입이 되지 않아 계속 null이 들어오는 문제를 마지막 요청 기한 내에 해결하지 못했습니다. 😥
주석 처리하고 리뷰 요청 보냅니다..!!
@Component로 등록된 클래스이고, new 생성자를 사용하고 있지도 않는데 왜 주입이 안되는지 아직 파악이 잘 되지 않았습니다..
디버깅해보며 원인을 파악해야 할 것 같습니다.

aegis1920 replied:
- 일단 첫째로 `Authorization` 헤더에 encoded 값을 그대로 전달해주고 있습니다. `basic ` 문자열도 포함해서 전달이 필요합니다.
- `@ExtendWith(MockitoExtension.class)`를 쓸 필요 없이 org.springframework.mock.web 패키지에 있는 MockHttpServletRequest을 사용하셔도 좋을 것 같습니다. (여기에 `addHeader()`를 통해 헤더만 추가하는 방식, 이렇게 하면 mockito를 의존하지 않고도 테스트 가능합니다)
- BasicAuthorizationExtractor이 주입되지 않는 이유는 TestContext에 해당 BasicAuthorizationExtractor Bean이 있지 않기 때문입니다. (TestContextManager -> TestContext -> applicationContext에서 basicAuthorizationExtractor을 getBean 해보시면 알 수 있습니다)
- TestContext에 해당 Bean을 넣어주기 위해선 아래와 같은 방법이 있습니다.

```java
// 1
@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = BasicAuthorizationExtractor.class)
class BasicAuthorizationExtractorTest {

// 2
@SpringBootTest(classes = BasicAuthorizationExtractor.class)
class BasicAuthorizationExtractorTest {

// 3
@ExtendWith(SpringExtension.class)
@Import(BasicAuthorizationExtractor.class)
@TestConfiguration
class BasicAuthorizationExtractorTest {
```

yoondgu replied:
> 일단 첫째로 Authorization 헤더에 encoded 값을 그대로 전달해주고 있습니다. basic  문자열도 포함해서 전달이 필요합니다.

헉 이건 스스로도 먼저 발견했어야 하는 실수인데 그러지 못했네요. ㅜㅜ 감사합니다..!

또, Test에서 ApplicationContext를 생성 한다고 정의된 모든 빈을 사용할 수 있는 게 아니라, 등록하고 로드할 수 있도록 위와 같은 방법으로 지정해주어야 한다는 사실을 알게 되었습니다!
알려주신 방법을 통해 해결했고, MockHttpServletRequest를 적용했습니다!!

yoondgu commented:
안녕하세요 빙봉!! 남겨주신 피드백 목록을 README.md에 함께 정리, 커밋하며 추가 리팩터링 진행하였습니다.
마감 기한을 코앞에 두고 요청드렸네요.. ㅎㅎ
이번 리뷰도 잘 부탁드립니다 🙇‍♀️🙇‍♀️

aegis1920 commented on src/main/java/cart/exception/ExceptionAdvice.java:
기본적인 예외들에 대해 Handle되어있는 ResponseEntityExceptionHandler를 활용해도 좋을 것 같습니다!

yoondgu replied:
좋은 것 같아요!! 이를 상속하도록 만들고, MethodArgumentNotValidException는 ResponseEntityExceptionHandler에서 처리해주므로 해당 예외를 처리하는 메서드는 삭제하였습니다.

aegis1920 reviewed (changes requested):
안녕하세요 도이! 빙봉입니다 🙂 

질문에 대한 답변 및 코멘트 남겼습니다. 이미 요구사항을 충분히 만족했지만 테스트 관련해서 질문주신 부분을 반영하시면 도이에게도 좋을 것 같아서 다시 한 번 요청드려요! 다시 리뷰요청 보내주시면 머지할테니 부담 갖지 않으셨으면 합니다ㅎㅎ

궁금한 점이 있다면 언제든 편하게 DM 주세요!

yoondgu committed:
fix: BasicAuthorizationExtractorTest 오류 해결

yoondgu committed:
refactor: 예외 처리 ResponseEntityExceptionHandler 활용

yoondgu commented:
안녕하세요 빙봉! 테스트 관련 부분, 추가 피드백 반영하였습니다.

마지막까지 어려움을 겪는 부분 도와주셔서 정말 감사합니다!!
빙봉과 코드리뷰하며 정말 많은 것을 배운 것 같습니다 ㅎ.ㅎ 그동안 감사했습니다!
남은 한 주도 즐겁게 보내시길 바랄게요!

aegis1920 reviewed (approved):
안녕하세요 도이! 빙봉입니다 🙂 

장바구니 미션하느라 고생 많으셨습니다! 끝까지 피드백 반영해주셔서 감사해요 💯 
장바구니 미션은 여기서 머지하겠습니다

지하철 미션도 파이팅하시길 바라요! 수고하셨습니다!
